demo/dynamic-fraud-detection/flink-job/src/main/java/com/ververica/field/dynamicrules/functions/DynamicRuleFunction.java

```diff
@@ -0,0 +1,264 @@
+
+  private void handleControlCommand(
+      Rule command, BroadcastState<Integer, Rule> rulesState, Context ctx) throws Exception {
+    ControlType controlType = command.getControlType();
+    switch (controlType) {
+      case EXPORT_RULES_CURRENT:
+        for (Map.Entry<Integer, Rule> entry : rulesState.entries()) {
+          ctx.output(Descriptors.currentRulesSinkTag, entry.getValue());
+        }
+        break;
+      case CLEAR_STATE_ALL:
+        //        rulesState.put(CLEAR_STATE_COMMAND_KEY, command);
+        ctx.applyToKeyedState(windowStateDescriptor, (key, state) -> state.clear());
+        break;
+      case CLEAR_STATE_ALL_STOP:
+        rulesState.remove(CLEAR_STATE_COMMAND_KEY);
+        break;
+      case DELETE_RULES_ALL:
+        Iterator<Entry<Integer, Rule>> entriesIterator = rulesState.iterator();
+        while (entriesIterator.hasNext()) {
+          Entry<Integer, Rule> ruleEntry = entriesIterator.next();
+          rulesState.remove(ruleEntry.getKey());
+          log.info("Removed Rule {}", ruleEntry.getValue());
+        }
+        break;
+    }
+  }
+
+  private boolean isStateValueInWindow(
+      Long stateEventTime, Long windowStartForEvent, long currentEventTime) {
+    return stateEventTime >= windowStartForEvent && stateEventTime <= currentEventTime;
+  }
+
+  private void aggregateValuesInState(
+      Long stateEventTime, SimpleAccumulator<BigDecimal> aggregator, Rule rule) throws Exception {
+    Set<Transaction> inWindow = windowState.get(stateEventTime);
+    if (COUNT.equals(rule.getAggregateFieldName())
+        || COUNT_WITH_RESET.equals(rule.getAggregateFieldName())) {
+      for (Transaction event : inWindow) {
+        aggregator.add(BigDecimal.ONE);
+      }
+    } else {
+      for (Transaction event : inWindow) {
+        BigDecimal aggregatedValue =
+            FieldsExtractor.getBigDecimalByName(
+                rule.getAggregateFieldName(),
+                event); // Should be double or BigDecimal in the first place
+        aggregator.add(aggregatedValue);
+      }
+    }
+  }
+
+  private boolean noRuleAvailable(Rule rule) {
+    // This happens if the BroadcastState in this CoProcessFunction was updated before it was
+    // updated in `DynamicKeyFunction`
+    // TODO Maybe we should consider sending the whole Rule Object with each event to avoid this
+    // race condition.
+    if (rule == null) {
+      log.info(
+          "Rule for ID {} was not found. Ignoring rule for this transaction. This should only happen close to an update to a rule.");
+      return true;
+    }
+    return false;
+  }
+
+  private void updateWidestWindowRule(Rule rule, BroadcastState<Integer, Rule> broadcastState)
+      throws Exception {
+    Rule widestWindowRule = broadcastState.get(WIDEST_RULE_KEY);
+    if (widestWindowRule != null && widestWindowRule.getRuleState() == Rule.RuleState.ACTIVE) {
+      if (widestWindowRule == null || widestWindowRule.getWindowMillis() < rule.getWindowMillis()) {
```

Comment:
You already made this null check on the outer conditional. 

```suggestion
      if (widestWindowRule.getWindowMillis() < rule.getWindowMillis()) {
```